### PR TITLE
micro-ROS User APIs [Parameters]

### PR DESCRIPTION
**Pull request by dcoello-eProsima**
Signed-off-by: David Coello Pulido <davidcoello@eprosima.com>

### CODE REVIEW

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        if (RCL_RET_OK != rc)
+        {
+            ... // Handle error
+            return -1;
+        }
+
+Memory requirements
+^^^^^^^^^^^^^^^^^^^
+
+The parameter server uses five services and an optional publisher. These need to be taken into account on the rmw-microxrcedds package memory configuration:
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:39]:
```suggestion
The parameter server uses five services and an optional publisher. These need to be taken into account on the `rmw_microxrcedds` package memory configuration:

.. TODO(pgarrido): Add link to memory conf tutorial when ready


```

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+.. code-block:: c
+
+    // Executor init example with the minimum RCLC executor handles required
+    rclc_executor_t executor = rclc_executor_get_zero_initialized_executor();
+    rc = rclc_executor_init(
+        &executor, &support.context,
+        RCLC_PARAMETER_EXECUTOR_HANDLES_NUMBER, &allocator);
+
+Humble: the variable is ``RCLC_PARAMETER_EXECUTOR_HANDLES_NUMBER`` has been renamed to ``RCLC_EXECUTOR_PARAMETER_SERVER_HANDLES``.
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:68]: We are supporting only humble, rephrase this paragraph and the previous so we use `RCLC_EXECUTOR_PARAMETER_SERVER_HANDLES`

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+        // Add parameter to the server
+        rcl_ret_t rc = rclc_add_parameter(&param_server, parameter_name, RCLC_PARAMETER_DOUBLE);
+
+        // Set parameter value
+        rc = rclc_parameter_set_double(&param_server, parameter_name, param_value);
+
+        // Get parameter value on param_value
+        rc = rclc_parameter_get_double(&param_server, parameter_name, &param_value);
+
+Max name size is controlled by the compile-time option `RCLC_PARAMETER_MAX_STRING_LENGTH`, default value is 50.
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:131]:
```suggestion
The parameter string name size is controlled by the compile-time option ``RCLC_PARAMETER_MAX_STRING_LENGTH``, the default value is 50.
```

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+.. code-block:: c
+
+    // Delete parameter server
+    rclc_parameter_server_fini(&param_server, &node);
+
+This will delete any automatically created infrastructure on the agent (if possible) and deallocate used memory on the parameter server side.
+
+micro-ROS Humble
+----------------
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:146]: We are only supporting humble

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+
+Initialization options
+^^^^^^^^^^^^^^^^^^^^^^
+- Custom options:
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:150]:
```suggestion
A parameter server can be configured at configuration time, the following options can be adjusted:
```

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+    The following options can be configured:
+
+    - notify_changed_over_dds: Publish parameter e  vents to other ROS 2 nodes as well.
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:154]:
```suggestion
    - notify_changed_over_dds: Publish parameter events to other ROS 2 nodes as well.
```

Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:154]:
```suggestion
    - ``notify_changed_over_dds``: Publish parameter e  vents to other ROS 2 nodes as well.
```

This in all elements

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst (hunk 5 above)
Comment (pablogs9) [reply docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:150]: decrease the indent level in all this list

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+        // Initialize parameter server with configured options
+        rcl_ret_t rc = rclc_parameter_server_init_with_option(&param_server, &node, &options);
+
+        if (RCL_RET_OK != rc)
+        {
+            ...     // Handle error
+            return -1;
+        }
+
+- Low memory mode:
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:180]: Make this a subsection and explain that we have a low memory mode with the following functionality

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+Parameters modifications are disabled while the callback ``on_parameter_changed`` is executed, causing the following methods to return ``RCLC_PARAMETER_DISABLED_ON_CALLBACK`` if they are invoked:
+
+- rclc_add_parameter
+- rclc_delete_parameter
+- rclc_parameter_set_bool
+- rclc_parameter_set_int
+- rclc_parameter_set_double
+- rclc_set_parameter_read_only
+- rclc_add_parameter_constraint_double
+- rclc_add_parameter_constraint_integer
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:267]: All those with the `` to use code block

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+    int64_t int_to = 20;
+    uint64_t int_step = 2;
+    rclc_add_parameter_constraint_integer(&param_server, "param2", int_from, int_to, int_step);
+
+    double double_from = -0.5;
+    double double_to = 0.5;
+    double double_step = 0.01;
+    rclc_add_parameter_constraint_double(&param_server, "param3", double_from, double_to, double_step);
+
+`This constrains will not be applied by the parameter server, leaving values filtering to the user callback`.
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:334]: This as a note

**Review by pablogs9 — CHANGES_REQUESTED** [review]
In general remove all those paragraphs as bullet point lists

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+        ...     // Handle error
+        return -1;
+    }
+
+Low memory mode
+^^^^^^^^^^^^^^^
+
+There is a low memory mode that ports the parameter functionality to memory constrained devices. The following constrains are applied:
+
+- Request size limited to one parameter on Set, Get, Get types and Describe services.
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:181]:
```suggestion
- Request size limited to one parameter on Set, Get, Get types and Describe operations.
```

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+- List parameter request has no prefixes enabled nor depth.
+- Parameter description strings not allowed, ``rclc_add_parameter_description`` is disabled.
+- Memory benchmark results on ``STM32F4`` for 7 parameters with ``RCLC_PARAMETER_MAX_STRING_LENGTH = 50`` and ``notify_changed_over_dds = true``:
+- Full mode: 11736 B
+- Low memory mode: 4160 B
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:186]:
```suggestion
.. note::

    Using low memory mode in a STM32F4 with  7 parameters with ``RCLC_PARAMETER_MAX_STRING_LENGTH =         50`` and ``notify_changed_over_dds = true`` the memory usage drops from 11.7 kB to 4.1 kB.
```

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+.. code-block:: c
+
+    rclc_delete_parameter(&param_server, "param2");
+
+For external delete requests, the server callback will be executed, allowing the node to reject the operation.
+
+Parameters description
+^^^^^^^^^^^^^^^^^^^^^^
+
+- Parameter description Adds a description of a parameter and its constrains, which will be returned on a describe parameter requests:
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:306]:
```suggestion
- Parameter description: Adds a description of a parameter and its constraints, which will be returned on a describe parameter request:
```

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+
+    .. code-block:: c
+
+        rclc_add_parameter_description(&param_server, "param2", "Second parameter", "Only even numbers");
+
+    The maximum string size is controlled by the compilation time option ``RCLC_PARAMETER_MAX_STRING_LENGTH``, default value is 50.
+
+- Parameter constraints Informative numeric constraints can be added to int and double parameters, returning this values on describe parameter requests:
```
Comment (pablogs9) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:314]:
```suggestion
- Parameter constraints: Informative numeric constraints that can be added to int and double parameters, returning these values on describe parameter requests:
```

**Comment by pablogs9** [comment]
LGTM but, please fix the CI:

```
./rst/tutorials/micro/user_api/user_api_parameter_server.rst:184: D000 Bullet list ends without a blank line; unexpected unindent.
```

File: docs/rst/spelling_wordlist.txt [hunk 14]
```diff
@@ -64,3 +64,4 @@ Vulcanexus
 turtlesim
 webpage
 ws
+unexisting
```
Comment (pablogs9) [inline docs/rst/spelling_wordlist.txt:67]: order this alphabetically

File: docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+    double double_from = -0.5;
+    double double_to = 0.5;
+    double double_step = 0.01;
+    rclc_add_parameter_constraint_double(&param_server, "param3", double_from, double_to, double_step);
+
+.. note::
+
+    This constrains will not be applied by the parameter server, leaving values filtering to the user callback.
+
+- Read-only parameters: The new API offers a read-only flag. This flag blocks parameter changes from external clients, but allows changes on the server side:
```
Comment (Acuadros95) [inline docs/rst/tutorials/micro/user_api/user_api_parameter_server.rst:336]:
```suggestion
- Read-only parameters: This flag blocks parameter changes from external clients, but allows changes on the server side:
```

**Review by Acuadros95 — APPROVED** [review]

**Pull request merged**